overlord/snapstate/aliasesv2.go

```diff
@@ -0,0 +1,383 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+// AliasState describes the state of an alias in the context of a snap.
+// aliases-v2 top state entry is a snapName -> alias -> AliasState map.
+type AliasState struct {
+	Status        string `json:"status"` // one of: auto,disabled,manual,overridden
+	Target        string `json:"target"`
+	AutoTargetBak string `json:"auto-target,omitempty"`
+}
+
+func (as *AliasState) Enabled() bool {
+	switch as.Status {
+	case "auto", "manual", "overridden":
+		return true
+	}
+	return false
+}
+
+func (as *AliasState) Manual() bool {
+	switch as.Status {
+	case "manual", "overridden":
+		return true
+	}
+	return false
+}
+
+func (as *AliasState) AutoTarget() string {
+	switch as.Status {
+	case "manual":
+		return ""
+	case "overridden":
+		return as.AutoTargetBak
+	default:
+		return as.Target
+	}
+}
+
+/*
+   There are two kinds of aliases:
+
+   * automatic aliases listed with their target application in the
+     snap-declaration of the snap (states: auto,disabled,overridden)
+
+   * manual aliases setup with "snap alias SNAP.APP ALIAS" (states:
+     manual,overridden)
+
+   Further
+
+   * all automatic aliases of a snap are either enabled (state: auto)
+     or disabled together
+
+   * disabling a manual alias removes it from disk and state (for
+     simplicity there is no disabled state for manual aliases)
+
+   * therefore enabled automatic aliases and manual aliases never mix
+     for a snap (no snap with mixed auto and manual/overridden states)
+
+   * a snap with manual aliases can only have disabled or overridden
+     automatic aliases, the latter means a manual alias overlaid on
+     top of what is declared by the snap as an automatic alias
+
+   Given that aliases are shared and can conflict between snaps and
+   (because of --prefer/prefer) some operations might need to touch
+   implicitly the aliases of other snaps, the execution of all tasks
+   touching alias states or aliases on disk are serialized, see
+   SnapManager.blockedTask
+*/
+
+// TODO: helper from snap
+func composeTarget(snapName, targetApp string) string {
+	if targetApp == snapName {
+		return targetApp
+	}
+	return fmt.Sprintf("%s.%s", snapName, targetApp)
+}
+
+// applyAliasChange applies the necessary changes to aliases on disk to go from prevStates to newStates for the aliases of snapName. It assumes that conflicts have already been checked.
```

Comment:
can you wrap this comment so it at least fits in github's diff view without panning? :-)